File: src/dotnet/Ansight.Tools.Reflection/ListReflectionRootsTool.cs
namespace Ansight.Tools.Reflection;

public sealed class ListReflectionRootsTool : ITool
{
    private readonly ReflectionToolsOptions options;

    public ListReflectionRootsTool(ReflectionToolsOptions? options = null)
    {
        this.options = options ?? ReflectionToolsOptions.Default;
    }

    public string Category => "reflect";

    public ToolScope Scope => ToolScope.Read;

    public string Id => ReflectionToolIds.ListRoots;

    public string Name => "List Reflection Roots";

    public string Description => "Lists the registered live object roots available for reflection tools.";

    public string Keywords => "reflection runtime inspect roots objects";

    public ToolSchema ArgumentsSchema => ReflectionToolSchemas.ListRootsArguments;

    public ToolSchema ResultSchema => ReflectionToolSchemas.ListRootsResult;

    public ToolSecurity Security => ReflectionToolSecurityProfiles.ListRoots;

    public Task<ToolResult> Execute(IReadOnlyDictionary<string, string> arguments)
    {
        ArgumentNullException.ThrowIfNull(arguments);

        try
        {
            return Task.FromResult(ToolResult.Success(ReflectionSupport.ListRoots(options)));
        }
        catch (Exception exception)
        {
            return Task.FromResult(ToolResult.Failure(exception.Message, errorCode: "reflect_list_roots_failed"));
        }
    }
}

File: src/ios/Sources/AnsightToolsReflection/ListReflectionRootsTool.swift
import AnsightCore
import Foundation

public final class ListReflectionRootsTool: AnsightTool {
    private let runtime: AnsightRuntime
    private let options: AnsightReflectionToolsOptions

    public init(
        options: AnsightReflectionToolsOptions = .default,
        runtime: AnsightRuntime = .shared
    ) {
        self.runtime = runtime
        self.options = options
    }

    public var descriptor: AnsightToolDescriptor {
        AnsightToolDescriptor(
            id: AnsightReflectionToolIds.listRoots,
            name: "List Reflection Roots",
            description: "Lists registered iOS runtime object roots available for reflection tools.",
            category: "reflect",
            scope: AnsightToolScope.read.rawValue,
            keywords: "reflection runtime inspect roots objects",
            security: AnsightReflectionToolSecurityProfiles.listRoots,
            argumentsSchema: AnsightReflectionToolSchemas.listRootsArguments,
            resultSchema: AnsightReflectionToolSchemas.listRootsResult
        )
    }

    public func execute(arguments: [String: String]) throws -> AnsightToolExecutionResult {
        .success(AnsightReflectionSupport.listRoots(runtime: runtime, options: options))
    }
}
